### PR TITLE
Add cmdstanr compatibility

### PR DESCRIPTION
**Pull request by n-kall**
The latests cmdstanr (not yet released, but on github) has functionality for `log_prob` and `unconstrain_pars` methods. So I adapted the `stanfit` method to one for `CmdStanFit`.

### CODE REVIEW

**Comment by topipa** [comment]
Thanks for opening this PR! I added you as a collaborator now in this repository. Let me know if you want me to review this before merging :slightly_smiling_face:

File: tests/testthat/test-moment-match-stanfit-analytical.R [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+mean_analytical_prior <- c(mu = mu_n, sigma_sq = sigma_sq_post_mean)
+sd_analytical_prior <- c(mu = mu_post_sd, sigma_sq = sqrt(sigma_sq_post_var))
+
+
+test_that("moment_match_stanfit matches analytical results (prior as proposal)", {
+
+  # ratio = jointlikelihood
+  joint_log_lik <- function(draws, fit, ...) {
+
+    cdraws <- constrain_draws_stanfit(fit, draws)
```
Comment (topipa) [inline tests/testthat/test-moment-match-stanfit-analytical.R]: I get a warning about this, should it be just constrain_draws?

File: tests/testthat/test-moment-match-cmdstan-analytical.R [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+mean_analytical_prior <- c(mu = mu_n, sigma_sq = sigma_sq_post_mean)
+sd_analytical_prior <- c(mu = mu_post_sd, sigma_sq = sqrt(sigma_sq_post_var))
+
+
+test_that("moment_match.CmdStanFit matches analytical results", {
+
+  # ratio = jointlikelihood
+  joint_log_lik <- function(draws, fit, ...) {
+
+    cdraws <- constrain_draws_CmdStanFit(fit, draws)
```
Comment (topipa) [inline tests/testthat/test-moment-match-cmdstan-analytical.R]: I get a warning about this, should it be just constrain_draws?

File: R/cmdstanr_functions.R [hunk 3]
```diff
@@ -0,0 +1,109 @@
+#' Generic importance weighted moment matching algorithm for `stanfit` objects.
+#' See additional arguments from `moment_match.matrix`
```
Comment (topipa) [inline R/cmdstanr_functions.R]: These should say `CmdStanFit`, right?

File: R/moment_match.R [hunk 4]
```diff
@@ -114,7 +114,8 @@ moment_match.matrix <- function(x,
     lw_psis <- suppressWarnings(loo::psis(lw))
     lw <- as.vector(weights(lw_psis))
     k <- lw_psis$diagnostics$pareto_k
-    if (is.infinite(k)) {
+    if (any(is.infinite(k))) {
+      print(exp(lw))
```
Comment (topipa) [inline R/moment_match.R]: Is this print intentional?

**Review by topipa — APPROVED** [review]
Overall this looks good to me :+1: Thanks for improving the structure and adding the workflows about linting and coverage etc.

I left a few small comments. The old tests in `test-moment-match-stanfit.R` can also be made to work with relatively small changes. I can clean them up and add them after this is merged, for now your analytical tests will do just fine.

**Comment by n-kall** [comment]
I haven't done much with this as I am waiting for the cmdstanr unconstrain_pars etc. interface to be finalised. I'll then get back to dealing with your comments. :)

**Comment by n-kall** [comment]
@topipa I think this is now in a good state to merge. Then we can think about a CRAN release. Would be great if you can take a look at the new tests and see if you think they make sense

File: R/cmdstanr_functions.R (hunk 3 above)
Comment (n-kall) [reply R/cmdstanr_functions.R]: fixed

File: tests/testthat/test-moment-match-stanfit-analytical.R (hunk 1 above)
Comment (n-kall) [reply tests/testthat/test-moment-match-stanfit-analytical.R]: Fixed

File: tests/testthat/test-moment-match-cmdstan-analytical.R (hunk 2 above)
Comment (n-kall) [reply tests/testthat/test-moment-match-cmdstan-analytical.R]: Fixed

File: R/moment_match.R (hunk 4 above)
Comment (n-kall) [reply R/moment_match.R]: Not intentional, it should be fixed now

**Comment by n-kall** [comment]
> Would be great if you can take a look at the new tests and see if you think they make sense

I just realised you already checked the tests last review, somehow I missed that. So I don't think anything major changed, just kept up to date with new methods in cmdstanr

**Comment by n-kall** [comment]
Currently the tests are failing on Windows, but perhaps we can deal with that after the merge

**Review by topipa — APPROVED** [review]
@n-kall still looks good. Feel free to merge this.

**Pull request merged**